### PR TITLE
feat: create a tool to be able to generate the AMD to ESM bridges automatically

### PR DESCRIPTION
**Pull request by izaera**
[ [https://issues.liferay.com/browse/IFI-2977](https://issues.liferay.com/browse/IFI-2977) ]

This modifies `npm-scripts` so that if a project declares `exports` in `npmscripts.config.js`, like this:

```javascript
module.exports = {
	build: {
		exports: {
			'react': 'react/index',
			'react-dom': 'react-dom/index',
		}
	},
};
```

The `npm-scripts` generate ESM modules that provide a bridge between AMD and ESM worlds, like this:

image

Where `react.js`, for example, contains:

```javascript
const amdModule = await new Promise((resolve, reject) => {
	Liferay.Loader.require(
		'liferay!frontend-js-react-web$react@16.12.0/index',
		resolve,
		reject
	);
});

const {
	Children,
	createRef,
	...
	version,
	__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
} = amdModule;

export {
	Children,
	createRef,
	...
	version,
	__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
};

export default amdModule;
```

We then combine these bridges with importmaps so that anyone can do this in an ESM module and it works without any transpiling or bundling:

```javascript
import react from 'react';
```

### CODE REVIEW

File: projects/npm-tools/packages/npm-scripts/src/utils/createExportsBridges.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+const {
+	addNamespace,
+	joinModuleName,
+	splitModuleName,
+} = require('@liferay/js-toolkit-core');
+const fs = require('fs');
+const path = require('path');
+const resolve = require('resolve');
+
+/* eslint-disable @liferay/no-dynamic-require */
```
Comment (bryceosterhaus) [inline projects/npm-tools/packages/npm-scripts/src/utils/createExportsBridges.js:15]: nit: I think this should technically be above the require statements. File-wide ignores should be right under the copyright

**Review by bryceosterhaus — APPROVED** [review]
LGTM! I like how simple the script seems

File: projects/npm-tools/packages/npm-scripts/src/utils/createExportsBridges.js (hunk 1 above)
Comment (izaera) [reply projects/npm-tools/packages/npm-scripts/src/utils/createExportsBridges.js]: done :+1:

**Comment by izaera** [comment]
Had to rebase and remove a dangling "WIP" commit :scream:

**Comment by izaera** [comment]
I'm merging this and will release a new version of npm-scripts so that the functionality can be tested on top of [the corresponding portal PR](https://github.com/brianchandotcom/liferay-portal/pull/112369)

**Pull request merged**